### PR TITLE
Only hash keys once

### PR DESCRIPTION
**Pull request by arthurprs**
Only hash keys once, by taking advantage of hashbrown raw entry api. This is roughly the same as #194.

### CODE REVIEW

**Comment by xacrimon** [comment]
Won't this break on resize AKA the reason the previous PR was reverted?

**Comment by arthurprs** [comment]
Maybe I'm missing something, but I don't know how that would break. Can you point me to prior discussion?

AFAICT the number of shards and the Hasher are fixed for the lifetime of the DashMap. Thus, the routing result should remain correct for the lifetime of DashMap and so is the hash passed down to the hashbrown shard.

**Comment by xacrimon** [comment]
lgtm

File: src/lib.rs [hunk 1]
```diff
@@ -1028,16 +1030,18 @@ impl<'a, K: 'a + Eq + Hash, V: 'a, S: 'a + BuildHasher + Clone> Map<'a, K, V, S>
...
 
-        if let Some((kptr, vptr)) = shard.get_key_value(key) {
+        if let RawEntryMut::Occupied(mut entry) =
+            shard.raw_entry_mut().from_key_hashed_nocheck(hash, key)
+        {
             unsafe {
-                let kptr: *const K = kptr;
-                let vptr: *mut V = vptr.as_ptr();
+                let kptr: *const K = entry.key();
+                let vptr: *mut V = entry.get_mut().as_ptr();
```
Comment (thalesfragoso) [inline src/lib.rs:1044]: I believe this violates stacked borrows, the `entry.get_mut()` invalidates the `*const` to the key. Inverting the order of the operations should be enough though.

**Comment by Turbo87** [comment]
👋  I'm not sure how exactly, but it looks like this change broke the crate download counting code on crates.io (see https://github.com/rust-lang/crates.io/issues/7039). Admittedly, we are us the `dashmap` API in some interesting ways, but I still didn't expect these changes to have such a big effect on us... 😅

**Comment by xacrimon** [comment]
hm, interesting. I'll yank and revert for now.

**Comment by xacrimon** [comment]
@Turbo87 Thanks for reporting. v5.5.2 is now out with this PR reverted. Since the earlier release has some other changes, would you guys mind updating to 5.5.2 and seeing if the issue is still resolved?

**Comment by arthurprs** [comment]
This is quite surprising. Seems like some bad interaction with dashmap raw-api.

**Comment by arthurprs** [comment]
The problem seems to be that crates.io is replacing the Hashmaps yielded by the raw-api :grimacing: which in turn causes them to get a different hasher from the Default::default.

 Specifically the `mem::take(..)`
https://github.com/rust-lang/crates.io/blob/5b4b6aae45241c648379ed968e8ed02642db805f/src/downloads_counter.rs#L80
and
 https://github.com/rust-lang/crates.io/blob/5b4b6aae45241c648379ed968e8ed02642db805f/src/downloads_counter.rs#L93C5-L93C63

I believe the issue would be prevented by replacing the hashmap with a new hashmap with the same hasher as before. This hasher can be acquired via `DashMap::hasher()`.

**Comment by Turbo87** [comment]
> Since the earlier release has some other changes, would you guys mind updating to 5.5.2 and seeing if the issue is still resolved?

just deployed with v5.5.3 and I can confirm that the issue is still resolved

**Pull request merged**